### ISSUE TITLE
Need GPU disabled in chromedp to run

### ISSUE THREAD

**Issue by Gregory-Ledray** [agnivade/wasmbrowsertest#10]
Could you include some examples please? 

### Background
I'm trying to get this working here:
https://gitlab.com/polyapp-open-source/build-wasm/-/merge_requests/1
#### Steps to reproduce
`git clone https://gitlab.com/polyapp-open-source/build-wasm`
`cd build-wasm`
`git checkout -b "1-create-and-run-integration-tests-which-require-a-browser" "origin/1-create-and-run-integration-tests-which-require-a-browser"`
`chmod +rwx ./test_integration.sh`
`./test_integration.sh`

And that last command hangs forever even if all tests in main_integration_test.go are commented out and I comment out everything in `func main()` in frontend/main.go

**Comment by agnivade** [comment]
The quickstart and the next section are all you need to get up and running. I am not sure what "examples" are you looking for. Could you clarify ?

**Comment by agnivade** [comment]
Your code is also slightly buggy. I had to install wasmbrowsertest manually. 

```
10:22:33-~/play/build-wasm$./test_integration.sh 
integration testing WASM code (GOOS=js; GOARCH=wasm)
exec: "wasmbrowsertest": executable file not found in $PATH
FAIL	gitlab.com/polyapp-open-source/build-wasm/cmd/frontend	0.000s
FAIL
integration testing everything else (GOOS=default; GOARCH=default)
ok  	gitlab.com/polyapp-open-source/build-wasm/cmd/backend	1.008s
?   	gitlab.com/polyapp-open-source/build-wasm/cmd/webserver	[no test files]
```

But after that, it works fine

```
integration testing WASM code (GOOS=js; GOARCH=wasm)
--- FAIL: TestHelloFromWASM (0.00s)
    main_integration_test.go:12: hi
    main_integration_test.go:14: hi
--- FAIL: TestSetupCallback (0.00s)
panic: syscall/js: call of Value.Call on null [recovered]
	panic: syscall/js: call of Value.Call on null

goroutine 9 [running]:
testing.tRunner.func1(0x470200)
	/usr/local/go/src/testing/testing.go:874 +0x38
panic(0x2b7c0, 0x40a1a0)
	/usr/local/go/src/runtime/panic.go:679 +0x20
syscall/js.Value.Call(0x7ff8000000000002, 0x56231, 0x10, 0x437748, 0x2, 0x2, 0x7ff8000000000002)
	/usr/local/go/src/syscall/js/js.go:318 +0x7f
gitlab.com/polyapp-open-source/build-wasm/cmd/frontend.SetupCallback()
	/home/agniva/play/build-wasm/cmd/frontend/main.go:41 +0x6
gitlab.com/polyapp-open-source/build-wasm/cmd/frontend.TestSetupCallback(0x470200)
	/home/agniva/play/build-wasm/cmd/frontend/main_integration_test.go:20 +0x1
testing.tRunner(0x470200, 0x5dbe8)
	/usr/local/go/src/testing/testing.go:909 +0xe
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x2d
FAIL	gitlab.com/polyapp-open-source/build-wasm/cmd/frontend	0.233s
FAIL
integration testing everything else (GOOS=default; GOARCH=default)
ok  	gitlab.com/polyapp-open-source/build-wasm/cmd/backend	(cached)
?   	gitlab.com/polyapp-open-source/build-wasm/cmd/webserver	[no test files]
```

The panic is something you need to fix. But it does not hang. In general, I don't see any issue with wasmbrowsertest here.

**Comment by Gregory-Ledray** [comment]
I think I figured out what's going on.

I cloned wasmbrowsertest and tried the following:
1. GOOS=js GOARCH=wasm go build -o wasmbrowsertest
2. ./wasmbrowsertest frontend

Output:
> -bash: ./wasmbrowsertest: cannot execute binary file: Exec format error

I then tried the following:
1. go build -o wasmbrowsertestDefault
2. ./wasmbrowsertestDefault frontend

Output:
> 
nothing, because it hangs

I then tried:
1. go build -o frontend.test /path/to/main/build-wasm/cmd/frontend

Output:
> go: directory /path/to/main/build-wasm/cmd/frontend outside available modules

To get around this for now, I created a new go mod for the frontend:
1. mkdir ~/separate_frontend && cd ~/separate_frontend
2. go mod init gitlab.com/polyapp-open-source/separate-frontend
3. GOOS=js GOARCH=wasm go test -c
4. ~/wasmbrowsertest/wasmbrowsertestDefault separate_frontend.test
> [hangs]
5. ~/wasmbrowsertest/wasmbrowsertest separate_frontend.test
> cannot execute binary fine: Exec format error

So I was still at square 1. Assuming that I was supposed to be compiling with the default GOOS/GOARCH, I continued by adding some print statements and found that wasmbrowsertest was executing, but this line was hanging:
`err = chromedp.Run(ctx, tasks...)`

Removing that line caused the binary to finish, so I did some investigating (https://github.com/chromedp/chromedp/issues/297) and found adding this to be helpful:
```
if os.Getenv("WASM_HEADLESS") == "off" {
    // existing code
} else {
opts := append(chromedp.DefaultExecAllocatorOptions[:],
                        chromedp.DisableGPU,
                        chromedp.NoSandbox,
                )
                logger.Printf("%v", chromedp.DefaultExecAllocatorOptions[:])
                var cancel context.CancelFunc
                allocCtx, cancel = chromedp.NewExecAllocator(context.Background(), opts...)
                defer cancel()
}
```

After that it started to work for me (aka panic as expected):
```
integration testing WASM code GOOS=js; GOARCH=wasm
--- FAIL: TestSetupCallback (0.00s)
panic: syscall/js: call of Value.Call on null [recovered]
        panic: syscall/js: call of Value.Call on null

goroutine 9 [running]:
testing.tRunner.func1(0x46e200)
        /usr/local/go/src/testing/testing.go:874 +0x38
panic(0x2b7c0, 0x40a0e0)
        /usr/local/go/src/runtime/panic.go:679 +0x20
syscall/js.Value.Call(0x7ff8000000000002, 0x5622f, 0x10, 0x449f18, 0x2, 0x2, 0x7ff8000000000002)
        /usr/local/go/src/syscall/js/js.go:318 +0x7f
gitlab.com/polyapp-open-source/build-wasm/cmd/frontend.SetupCallback(0x92dda80001ecac5, 0x5e387c65)
        /mnt/c/Users/gledr/Polyapp_Apps/gocode/src/gitlab.com/polyapp-open-source/build-wasm/cmd/frontend/main.go:60 +0xd
gitlab.com/polyapp-open-source/build-wasm/cmd/frontend.TestSetupCallback(0x46e200)
        /mnt/c/Users/gledr/Polyapp_Apps/gocode/src/gitlab.com/polyapp-open-source/build-wasm/cmd/frontend/main_integration_test.
go:21 +0x2
testing.tRunner(0x46e200, 0x5dbf8)
        /usr/local/go/src/testing/testing.go:909 +0xe
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:960 +0x2d
FAIL    gitlab.com/polyapp-open-source/build-wasm/cmd/frontend  1.771s
FAIL
integration testing everything else GOOS=default; GOARCH=default
ok      gitlab.com/polyapp-open-source/build-wasm/cmd/backend   (cached)
?       gitlab.com/polyapp-open-source/build-wasm/cmd/webserver [no test files]

```

Interestingly, the code also works if chromedp.NoSandbox is not set as well, but I'd like to keep it in there to make sure it's aligned with what's needed for docker containers.

from https://github.com/chromedp/chromedp/issues/297#issuecomment-485892406
> Aside from that, all uses of headless-shell within a Docker image should use NoSandbox, as clearly required by Chrome. I'm not sure if chromedp should do this by default; we don't want to enforce it for all users, and detecting whether we're root and running headless-shell at run-time could be quite confusing to users.

So basically, if you don't mind I'd like to add that 'else' to this repo.

**Comment by agnivade** [comment]
Is there any documentation which explains why disabling GPU is needed inside a docker container ?

@mvdan any thoughts on this ? If `DisableGPU` is always needed, should it be included in the `DefaultExecAllocatorOptions` ?

**Comment by mvdan** [comment]
I'm not super experienced with running wasm via chromedp with a headless browser. All I know is that a headless browser with gpu capabilities is still useful, even if the physical hardware has no GPU - Chrome then uses [swiftshader](https://github.com/google/swiftshader) to implement webGL and others.

So, to directly answer your question, I don't think `DisableGPU` should be a default option. If someone uses a custom Chromium build without swiftshader, or doesn't need webGL, or somehow runs into errors like the above (though this should be investigated), they can disable the GPU.

**Comment by agnivade** [comment]
Thanks ! I will look around for more information about `DisableGPU`. I would want to enable it on some condition rather than unconditionally. And `NoSandbox` seems to be already handled with this [commit](https://github.com/chromedp/chromedp/commit/b053e659b88ccf63093375a46411f7a3d82eec0a) so I don't believe we need it.

**Comment by mvdan** [comment]
Yep, anything that chromedp can detect on its own should be handled automatically. Your use case is a bit different, because you know you'll only be loading wasm - if the use case doesn't include graphics, using `DisableGPU` by default sounds fine to me. `chromedp` is a more generic library, so it doesn't fit there as a default.

**Comment by agnivade** [comment]
Ah some interesting information- https://developers.google.com/web/updates/2017/04/headless-chrome#faq. And the last comment on the [crbug](https://bugs.chromium.org/p/chromium/issues/detail?id=737678#c10) does mention that this flag still seems to be necessary when running Linux inside a container.

**Comment by mvdan** [comment]
Hm, I don't follow. mesa and other GL libraries shouldn't be necessary if swiftshader is bundled with Chrome.

For example, this passes for me - it's running all of the chromedp tests inside a container with a headless build of Chrome. Note that `chromedp_test.go` does set `DisableGPU` for the options used for the tests, but commenting out the line doesn't break the tests.

So I don't think those links are up to date. Today, `--disable-gpu` is not required to get either containers or root working. But maybe it's still necessary for wasm to work well; I don't have experience there.

**Comment by agnivade** [comment]
Interesting .. @Gregory-Ledray can you shed some more light on your environment ? 

- your host environment
- docker environment
- Google chrome version
- Go version

**Comment by mvdan** [comment]
Also clarify what docker image you're using, and how you're running the container. As you can see above, my steps with a specific docker image work. That image contains a headless chromium build with swiftshader included. If swiftshader weren't included, `--disable-gpu` would probably be required.

**Comment by agnivade** [comment]
FWIW, I tested by manually installing google-chrome-stable and all its dependencies on a "golang:1.13" image and it works fine.

**Comment by Gregory-Ledray** [comment]
After poking around a bit more, reading what you guys did, and running some stuff, this looks like a very specific WSL issue. If I try to run this library with Windows it pops up with MINGW64 which works; if I run commands directly in Windows the GOOS / GOARCH isn't supported; if I run commands directly in WSL I need to have the change I put in for the code to work.

I'd guess this is the problem I'm running in to: https://github.com/microsoft/WSL/issues/3749

I'm currently trying a fresh install of Cygwin to see what happens.

PR-wise, perhaps there's a flag which can be checked to disable GPUs in WSL (I'll look into that in a bit).

**Comment by agnivade** [comment]
Are you using docker containers inside WSL ? Isn't that like a layer on a layer ? 

>  if I run commands directly in Windows the GOOS / GOARCH isn't supported; 

I am not sure why. wasm is supported on any underlying OS.

**Comment by Gregory-Ledray** [comment]
> Are you using docker containers inside WSL? Isn't that like a layer on a layer ? 

I'm not using Docker in WSL. Taking a step back from this issue, in https://gitlab.com/polyapp-open-source/build-wasm I'm trying to write some scripts to build, test, and deploy WASM code to websites. At the end of the project I want the script to work in several dev environments - WSL, MINGW, Windows, Linux - and be deployable to both those environments and several other environments - Google App Engine, Windows Server, and Go docker images (probably shoved into Kubernetes). To ensure a given change doesn't break any of those environments I'll probably use Docker images to regression test the test, build, run, and integration testing scripts. One of those images will have to be WSL.

Except as I was dev'ing the integration testing scripts I happened to get *super unlucky* because I'm currently dev'ing in the one dev environment I want to support which didn't work immediately. I then opened this issue.

> ... wasm is supported on any underlying OS.

My poor understanding of Windows scripting is now apparent to me. My tests do work (well, they panic as expected) on GOOS=js GOARCH=wasm directly in Windows with the current master branch if I type out each command individually.


Anyway, I'm still trying to figure out the best way to make the uname syscall to determine if an image is running WSL.

**Comment by agnivade** [comment]
> Anyway, I'm still trying to figure out the best way to make the uname syscall to determine if an image is running WSL.

Sounds great ! But syscalls would make it platform dependant. It would be preferable to keep the code as platform-independant as possible.

**Comment by Gregory-Ledray** [comment]
Yas! Thanks for working on my PR agnivade!
```
greg@LAPTOP-AH6IAHBO:~/test-build-wasm-6/build-wasm$ GOOS=js GOARCH=wasm go test -tags=integration -exec=/home/greg/test-build-w
asm-6/wasmbrowsertest/wasmbrowsertest ./...
--- FAIL: TestAlwaysFail (0.00s)
    main_integration_test.go:20: always fail
--- FAIL: TestSetupCallback (0.00s)
panic: syscall/js: call of Value.Call on null [recovered]
        panic: syscall/js: call of Value.Call on null

goroutine 10 [running]:
testing.tRunner.func1(0x46e300)
        /usr/local/go/src/testing/testing.go:874 +0x38
panic(0x2b7e0, 0x40a160)
        /usr/local/go/src/runtime/panic.go:679 +0x20
syscall/js.Value.Call(0x7ff8000000000002, 0x56268, 0x10, 0x437718, 0x2, 0x2, 0x7ff8000000000002)
        /usr/local/go/src/syscall/js/js.go:318 +0x7f
gitlab.com/polyapp-open-source/build-wasm/cmd/frontend.SetupCallback(0x29209080001ecce3, 0x5e401ca8)
        /home/greg/test-build-wasm-6/build-wasm/cmd/frontend/main.go:60 +0xd
gitlab.com/polyapp-open-source/build-wasm/cmd/frontend.TestSetupCallback(0x46e300)
        /home/greg/test-build-wasm-6/build-wasm/cmd/frontend/main_integration_test.go:26 +0x2
testing.tRunner(0x46e300, 0x5dc38)
        /usr/local/go/src/testing/testing.go:909 +0xe
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:960 +0x2d
FAIL    gitlab.com/polyapp-open-source/build-wasm/cmd/frontend  1.691s
FAIL
greg@LAPTOP-AH6IAHBO:~/test-build-wasm-6/build-wasm$

```
